src/Diagnostics.DataProviders/DataProviders/AppInsightsDataProvider.cs

```diff
@@ -0,0 +1,35 @@
+﻿using Diagnostics.DataProviders.DataProviderConfigurations;
+using Diagnostics.DataProviders.Interfaces;
+using Diagnostics.ModelsAndUtils.Models;
+using System.Data;
+using System.Threading.Tasks;
+
+namespace Diagnostics.DataProviders
+{
+    public class AppInsightsDataProvider : DiagnosticDataProvider, IDiagnosticDataProvider, IAppInsightsDataProvider
+    {
+        private readonly IAppInsightsClient _appInsightsClient;
+        private AppInsightsDataProviderConfiguration _configuration;
+
+        public AppInsightsDataProvider(OperationDataCache cache, AppInsightsDataProviderConfiguration configuration) : base(cache)
+        {
+            _configuration = configuration;
+            _appInsightsClient = new AppInsightsClient(_configuration);
+        }
+        public Task<bool> SetAppInsightsKey(string appId, string apiKey)
```

Comment:
Does this have to return a Task?

Reply:
Yes seems so. We expose all the functions of dataproviders from DataProviderLogDecorator.cs, and the wrapper MakeDependencyCall<T>(string requestId, Task<T> dataProviderTask, [CallerMemberName]string dataProviderOperation = "") inside only accepts a Task.

Reply:
Any method to a dataprovider were always though of as some kind of IO operation which would require awaiting (Executing Kusto Query, Observer API calls, GeoMaster API calls). When creating the decorator class, I didn't foresee anyone setting properties with it. But this works.

Reply:
There is one more problem we found in this design. You cannot have same method names in different Data Providers. Since Kusto Data Provider has an ExecuteQuery , so App Insights Data Provider cannot have method called ExecuteQuery, since at the end all methods needs to be redeclared in the decorator class..